storage/src/adb/current.rs

```diff
@@ -0,0 +1,555 @@
+            context,
+            bitmap_metadata_partition: config.bitmap_metadata_partition,
+        })
+    }
+
+    /// Get the number of operations that have been applied to this db, including those that are not
+    /// yet committed.
+    pub fn op_count(&self) -> u64 {
+        self.any.op_count()
+    }
+
+    /// Return the oldest location that remains readable & provable.
+    pub fn oldest_retained_loc(&self) -> Option<u64> {
+        self.any.oldest_retained_loc()
+    }
+
+    /// Get the value of `key` in the db, or None if it has no value.
+    pub async fn get(&self, key: &K) -> Result<Option<V>, Error> {
+        self.any.get(key).await
+    }
+
+    /// Updates `key` to have value `value`. If the key already has this same value, then this is a
+    /// no-op. The operation is reflected in the snapshot, but will be subject to rollback until the
+    /// next successful `commit`.
+    pub async fn update(
+        &mut self,
+        hasher: &mut H,
+        key: K,
+        value: V,
+    ) -> Result<UpdateResult, Error> {
+        let update_result = self.any.update(hasher, key, value).await?;
+        match update_result {
+            UpdateResult::NoOp => return Ok(update_result),
+            UpdateResult::Inserted(_) => (),
+            UpdateResult::Updated(old_loc, _) => {
+                self.is_active.set_bit(hasher, old_loc, false);
+            }
+        }
+        self.is_active.append(hasher, true);
+
+        Ok(update_result)
+    }
+
+    /// Delete `key` and its value from the db. Deleting a key that already has no value is a no-op.
+    /// The operation is reflected in the snapshot, but will be subject to rollback until the next
+    /// successful `commit`.
+    pub async fn delete(&mut self, hasher: &mut H, key: K) -> Result<(), Error> {
+        let Some(old_loc) = self.any.delete(hasher, key).await? else {
+            return Ok(());
+        };
+
+        self.is_active.append(hasher, false);
+        self.is_active.set_bit(hasher, old_loc, false);
+
+        Ok(())
+    }
+
+    /// Commit pending operations to the adb::any and sync it to disk.
+    async fn commit_ops(&mut self, hasher: &mut H) -> Result<(), Error> {
+        // Raise the inactivity floor by the # of uncommitted operations, plus 1 to account for the
+        // commit op that will be appended.
+        self.any
+            .raise_inactivity_floor(
+                hasher,
+                self.any.uncommitted_ops + 1,
+                Some(&mut self.is_active),
+            )
+            .await?;
+        self.any.uncommitted_ops = 0;
+        self.any.sync().await
+    }
+
+    /// Commit any pending operations to the db, ensuring they are persisted to disk &
+    /// recoverable upon return from this function.
+    pub async fn commit(&mut self, hasher: &mut H) -> Result<(), Error> {
+        self.commit_ops(hasher).await?;
+
+        // Prune inactive elements from the any db.
+        self.any.prune_inactive().await?;
```

Comment:
To confirm my understanding, do we just leave this outside of `commit_ops` for to make it easier to write failure tests?

If so, I'd recommend moving this into `commit_ops` and copying the required inner ops into those test functions instead ("keep the prod code clean").

Reply:
Yes though "commit" is the main function, which I broke apart into commit_ops to be used by failure simulation without having to duplicate code and risk the test code diverging from the what's going on in the prod flow. Since the impact on prod flow is pretty minor I thought the trade-off was worth it. You still prefer the copy?